test/py/test_migration.py

```diff
@@ -276,181 +314,193 @@ def test_migration_get_state():
+    data transfer size is 4 bytes. Here we test to check that a transfer of 8
+    bytes fails.
+    """
 
-    payload = vfio_user_mig_data(
-        argsz=len(vfio_user_mig_data()) + 8,
-        size=8
-    )
+    global read_data
 
-    # When we set up the tests at the top of this file we specify that the max
-    # data transfer size is 4 bytes. Here we test to check that a transfer of 8
-    # bytes fails.
+    transition_to_state(ctx, sock, VFIO_USER_DEVICE_STATE_RUNNING)
 
-    transition_to_state(VFIO_USER_DEVICE_STATE_PRE_COPY)
     read_data = bytes([1, 2, 3, 4, 5, 6, 7, 8])
-    msg(ctx, sock, VFIO_USER_MIG_DATA_READ, payload, expect=errno.EINVAL)
-
-
-def test_handle_mig_data_read_invalid_state():
-    global read_data
 
     payload = vfio_user_mig_data(
-        argsz=len(vfio_user_mig_data()) + 4,
-        size=4
+        argsz=len(vfio_user_mig_data()) + len(read_data),
+        size=len(read_data)
     )
 
-    data = bytes([1, 2, 3, 4])
-
-    transition_to_state(VFIO_USER_DEVICE_STATE_RUNNING)
-    read_data = data
-    msg(ctx, sock, VFIO_USER_MIG_DATA_READ, payload, expect=errno.EINVAL)
-
-    transition_to_state(VFIO_USER_DEVICE_STATE_STOP)
-    read_data = data
+    transition_to_state(ctx, sock, VFIO_USER_DEVICE_STATE_PRE_COPY)
     msg(ctx, sock, VFIO_USER_MIG_DATA_READ, payload, expect=errno.EINVAL)
 
 
 def test_handle_mig_data_read_failed_callback():
-    global fail_callbacks
+    transition_to_state(ctx, sock, VFIO_USER_DEVICE_STATE_PRE_COPY)
 
-    transition_to_state(VFIO_USER_DEVICE_STATE_PRE_COPY)
+    read_data = bytes([1, 2, 3, 4])
 
-    fail_callbacks = True
+    setup_fail_callbacks()
 
     payload = vfio_user_mig_data(
-        argsz=len(vfio_user_mig_data()) + 4,
-        size=4
+        argsz=len(vfio_user_mig_data()) + len(read_data),
+        size=len(read_data)
     )
 
     msg(ctx, sock, VFIO_USER_MIG_DATA_READ, payload, expect=errno.EINVAL)
+    assert c.get_errno() == errno.EINVAL
 
-    fail_callbacks = False
+    teardown_fail_callbacks()
 
 
 def test_handle_mig_data_read_short_write():
-    payload = struct.pack("I", 8)
+    data = bytes([1, 2, 3, 4])
 
-    msg(ctx, sock, VFIO_USER_MIG_DATA_READ, payload, expect=errno.EINVAL)
+    payload = vfio_user_mig_data(
+        argsz=len(vfio_user_mig_data()) + len(data),
+        size=len(data)
+    )
+
+    payload = bytes(payload)
+    assert len(payload) == 8
+
+    # don't send the last byte
+    msg(ctx, sock, VFIO_USER_MIG_DATA_READ, payload[0:7], expect=errno.EINVAL)
 
 
 def test_handle_mig_data_write():
+    data = bytes([1, 2, 3, 4])
+
     payload = vfio_user_mig_data(
-        argsz=len(vfio_user_mig_data()) + 4,
-        size=4
+        argsz=len(vfio_user_mig_data()) + len(data),
+        size=len(data)
     )
 
-    data = bytes([1, 2, 3, 4])
-
-    transition_to_state(VFIO_USER_DEVICE_STATE_RESUMING)
+    transition_to_state(ctx, sock, VFIO_USER_DEVICE_STATE_RESUMING)
```

Comment:
nit: since `ctx` and `sock` are global, can we have a helper method local to this file that passes them to `transition_to_state`?